src/main/java/info/skyblond/nekohit/neo/contract/WCAContract.java

```diff
@@ -309,8 +302,44 @@ public static void refund(String identifier, Hash160 buyer) throws Exception {
         }
 
         // update buyer info
-        ByteString buyerData = StdLib.serialize(buyerInfo);
-        wcaBuyerInfoMap.put(identifier, buyerData);
+        updateWCABuyerInfo(identifier, buyerInfo);
+    }
+
+    public static void cancelWCA(String identifier) throws Exception {
+        // get obj
+        WCABasicInfo basicInfo = getWCABasicInfo(identifier);
+        require(basicInfo != null, Messages.RECORD_NOT_FOUND);
+        List<WCAMilestone> milestones = getWCAMilestones(identifier);
+        require(milestones != null, Messages.BROKEN_RECORD);
+        WCABuyerInfo buyerInfo = getWCABuyerInfo(identifier);
+        require(buyerInfo != null, Messages.BROKEN_RECORD);
+        // check signature
+        require(Hash160.isValid(basicInfo.owner), Messages.INVALID_HASH160);
+        require(Runtime.checkWitness(basicInfo.owner) || basicInfo.owner == Runtime.getCallingScriptHash(), Messages.INVALID_SIGNATURE);
+        // check status
+        basicInfo.updateStatus(milestones);
+        switch (basicInfo.status) {
+            case 0:
+                // PENDING, nothing to do
+                break;
+            case 1:
+                // OPEN, refund to everyone
+                // to creator
+                transferTokenTo(basicInfo.owner, basicInfo.getTotalStake(), identifier);
+                // to buyers
+                Hash160[] buyers = buyerInfo.purchases.keys();
+                for (Hash160 buyer : buyers) {
+                    transferTokenTo(buyer, buyerInfo.purchases.get(buyer), identifier);
+                }
+                break;
+            default:
+                throw new Exception(Messages.INVALID_STATUS_ALLOW_PENDING_AND_OPEN);
```

Comment:
I don't think this is necessary, if the status is invalid (if this ever can happen) it's the wrong place to check for it. And if it is invalid or pending it should just result in doing nothing.

Reply:
This switch statement will handle different statuses. For `PENDING`, it's nothing to do, just delete the WCA. For `OPEN`, since someone might make the purchase, so we have to refund them. For the rest, like `ACTIVE` (passed the threshold) and `FINISHED` (tokens are transferred), the default branch will throw an exception and make this invoke failed.

Reply:
oh right, my bad. I think then it's better to use if-else statements.